MSAL/src/native_auth/network/MSALNativeAuthResponseErrorHandler.swift

```diff
@@ -0,0 +1,193 @@
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+import Foundation
+
+@_implementationOnly import MSAL_Private
+
+final class MSALNativeAuthResponseErrorHandler<T: Decodable & Error>: NSObject, MSIDHttpRequestErrorHandling {
+    private var customError: T?
+
+    // swiftlint:disable:next function_parameter_count
+    func handleError(
+        _ error: Error?,
+        httpResponse: HTTPURLResponse?,
+        data: Data?,
+        httpRequest: MSIDHttpRequestProtocol?,
+        responseSerializer: MSIDResponseSerialization?,
+        externalSSOContext ssoContext: MSIDExternalSSOContext?,
+        context: MSIDRequestContext?,
+        completionBlock: MSIDHttpRequestDidCompleteBlock?
+    ) {
+        guard let httpResponse = httpResponse else {
+            completionBlock?(nil, error)
+            return
+        }
+
+        if shouldRetry(httpResponse: httpResponse, httpRequest: httpRequest) {
+            retryRequest(httpRequest: httpRequest,
+                         context: context,
+                         completionBlock: completionBlock)
+            return
+        }
+
+        if httpResponse.statusCode == 400 || httpResponse.statusCode == 401 {
+            // PKeyAuth challenge
+            if let authValue = wwwAuthValue(httpResponse: httpResponse) {
+                handleAuthenticateHeader(wwwAuthValue: authValue,
+                                         httpRequest: httpRequest,
+                                         context: context,
+                                         ssoContext: ssoContext,
+                                         completionBlock: completionBlock)
+                return
+            }
+
+            handleAPIError(data: data, completionBlock: completionBlock)
+            return
+        }
+
+        handleHTTPError(httpResponse: httpResponse,
+                        context: context,
+                        completionBlock: completionBlock)
+    }
+
+    private func shouldRetry(httpResponse: HTTPURLResponse,
+                             httpRequest: MSIDHttpRequestProtocol?) -> Bool {
+        guard let httpRequest = httpRequest, httpRequest.retryCounter > 0 else {
+            return false
+        }
+        return httpResponse.statusCode >= 500 && httpResponse.statusCode <= 599
```

Comment:
I see some logic is common with MSIDAADRequestErrorHandler, should we refactor them to remove copy/paste?

Reply:
Thanks Danilo, it is a blocker.

Reply:
Thanks Sergei, we will work on it as soon as possible and we will add you as reviewer of the PR